lib/services/database/challenge_repository_service.dart

```diff
@@ -146,20 +146,25 @@ class ChallengeRepositoryService {
     final Iterable<PostIdFirestore> possiblePosts =
         await _inRangeUnsortedPosts(pos, excludedUser);
 
-    // The whereIn argument of the where method crashed
+    // The [whereIn] argument of the where method crashed
     // if the query is empty for the real firestore. This
     // cannot be tested with the mock firestore.
+    // (It is now fixed with the local filtering below, but
+    // we can keep it for the sake of optimization.)
     if (possiblePosts.isEmpty) return;
 
     final Iterable<String> possiblePostsStringIds =
         possiblePosts.map((post) => post.value);
 
     final pastChallengesCollectionRef = _pastChallengesRef(parentRef);
-    final alreadyDonePostsSnap = await pastChallengesCollectionRef
-        .where(FieldPath.documentId, whereIn: possiblePostsStringIds)
-        .get();
+    final alreadyDonePostsSnap = await pastChallengesCollectionRef.get();
 
+    // The [whereIn] argument of the where method crashed
+    // if the query has more than 30 elements for the real firestore.
+    // So we do the filtering locally rather than using the whereIn parameter
+    // of the where method.
     final alreadyDonePosts = alreadyDonePostsSnap.docs
+        .where((donePost) => possiblePostsStringIds.contains(donePost.id))
```

Comment:
Here I think we should avoid string comparison of IDs. It defeats the purposes of wrapping IDs in opaque types. (I am not really for wrapping IDs in the first place but as it is the decision when went with I think it is important to stick to it.)

Therefore, we should probably go with something like:

```dart
final alreadyDonePosts = alreadyDonePostsSnap.docs
    .map((post) => PostIdFirestore(value: post.id))
    .where((donePost) => possiblePosts.contains(donePost))
    .toSet();
```

That way, we don't accidentally make a mistake and compare two string IDs that do not reference the same types of objects (in theory).

On top of that, it is a bit more concise, avoids repetition and the need for the intermediary `possiblePostsStringIds` mapping iteration altogether.

Reply:
Good catch, you are completely right.

Fixed in: [56d1c86](https://github.com/ProximaEPFL/proxima/pull/261/commits/56d1c861c20f432d4b0e85eb50053e4ae16e5786)

Thank you.